src/browser-history/background/import-history.js

```diff
@@ -116,32 +117,66 @@ export default function importHistory({
     startTime,
     endTime,
 }={}) {
-    // Get the full history: both the historyItems and visitItems.
-    console.time('import history')
-    return getHistoryItems({startTime, endTime}).then(
-        getVisitItemsForHistoryItems
-    ).then(
-        // Convert everything to our data model
-        convertHistoryToPagesAndVisits
-    ).then(({pageDocs, visitDocs}) => {
-        // Mark and store the pages and visits.
-        let allDocs = pageDocs.concat(visitDocs)
-        // Mark each doc to remember it originated from this import action.
-        const importTimestamp = new Date().getTime()
-        allDocs = allDocs.map(doc => ({
-            ...doc,
-            importedFromBrowserHistory: importTimestamp,
-        }))
-        // Store them into the database. Already existing docs will simply be
-        // rejected, because their id (timestamp & history id) already exists.
-        return db.bulkDocs(allDocs)
-    }).then(() => {
-        console.timeEnd('import history')
-        console.time('rebuild search index')
-        return updatePageSearchIndex()
-    }).then(() => {
-        console.timeEnd('rebuild search index')
-    })
+    return function(dispatch, getState) {
+        dispatch(startLoadingHistory())
+        // Get the full history: both the historyItems and visitItems.
+        console.time('import history')
+        return getHistoryItems({startTime, endTime})
+            .then(data => {
+                if (getState().options.loadingHistoryStatus != 'pending') {
+                    // console.log('stopped on step 1')
+                    throw { status: 'canceled' }
```

Comment:
The only way I found yet how to stop Promise execution: check whether stop button was clicked after each async action within a promise